### ISSUE TITLE
ConstantReader.revive creates a broken reference for Curves.ease 

### ISSUE THREAD

**Issue by srawlins** [dart-lang/source_gen#533]
(User reported this at https://github.com/dart-lang/mockito/issues/425; mockito's code generation uses source_gen to revive parameter default values)

When trying to revive the default value of the `curve` parameter of [TabController.animateTo](https://api.flutter.dev/flutter/material/TabController/animateTo.html), `ConstantReader.revive` gives back `Cubic.ease`, which is not a real thing.

The default value of this parameter is [`Curves.ease`](https://api.flutter.dev/flutter/animation/Curves/ease-constant.html), declared as:

```dart
static const Cubic ease = Cubic(0.25, 0.1, 0.25, 1.0);
```

When I do some print-debugging, I see that `reviveInstance` gets all the way down to [checking fields](https://github.com/dart-lang/source_gen/blob/master/source_gen/lib/src/constants/revive.dart#L71), where `clazz` is `Cubic` and `e` is `Curves.ease`. So the result is `Revivable._(source: ..., accessor: 'Cubic.ease')` (again, not a real thing).

I think the fix might be as easy as adjusting the for-loop here to hold on to `t`, the type that `e` is found on, and using that rather than `clazz`.

As a stand-alone example, I have this:

```dart
class Foo {
  void f({Bar a = Bars.b}) {}
}
class Bars {
  static const Baz b = Baz();
}
class Bar {}
class Baz implements Bar {
  const Baz();
}
```

**Comment by kuhnroyal** [comment]
I think I run into this problem or something related when trying to mock a database generated with https://github.com/simolus3/moor

There is a `TableUpdateQuery` class here: https://github.com/simolus3/moor/blob/develop/moor/lib/src/runtime/api/stream_updates.dart#L122 with a factory
`const factory TableUpdateQuery.any() = AnyUpdateQuery;`.

The `AnyUpdateQuery` class is declared in another file here: https://github.com/simolus3/moor/blob/develop/moor/lib/src/runtime/executor/stream_queries.dart#L279

When the mock for the database is generated it contains this fragment for a method where the const factory is used as default value for a parameter:

```dart
import 'package:moor/src/runtime/api/runtime_api.dart' as _i29;
import 'package:moor/src/runtime/executor/stream_queries.dart' as _i32;

  @override
  _i35.Stream<Set<_i29.TableUpdate>> tableUpdates(
          [_i29.TableUpdateQuery? query = const _i32.TableUpdateQuery.any()]) =>
      (super.noSuchMethod(Invocation.method(#tableUpdates, [query]),
              returnValue: Stream<Set<_i29.TableUpdate>>.empty())
          as _i35.Stream<Set<_i29.TableUpdate>>);
```

The `_i32` import prefix is obviously wrong and doesn't point to the file containing the factory function but instead to the type that is being returned by the factory function. This should be `_i29` as well.

@simolus3 FYI

**Comment by kuhnroyal** [comment]
Actually meant to post this in the linked mockito issue :/
But may also be related to the `ConstantReader` used in mockito.

**Comment by natebosch** [comment]
I suspect we may still have some gaps here, but #546 got us a lot closer. I filed #547 to track the remaining bug.

**Comment by natebosch** [comment]
@srawlins - I just realized I failed to confirm whether the `Curves.ease` case is fixed or is still hit by #547

Do you have a repro set up to confirm whether this fixed it?

**Comment by srawlins** [comment]
Yes I just confirmed with a unit test that the correct text is revived, with this issue fixed. Thanks!

(This is with a minimal repro which aimed to emulate the Curves bug. I don't have a repro set up which uses Curves.)
